students/Mario_Alvarenga/lesson05/Mailroom_3.py

```diff
@@ -0,0 +1,204 @@
+        my_file = open(file_full_path,'w+')
+        my_file.write(message)
+        my_file.close()
+
+        #Log Path of where text file has exported to
+        print('Location of {dudes} file: {fpath}'.format(dudes = name,fpath = file_full_path))
+    print('\n\n')           ##Extra whitespace
+    return
+
+def f_donor_key(donor):
+    # Donor is a tuple of the form (name, total donation, number of donations, average donation)
+    # Sort by total donation
+    return donor[1]
+
+def f_generate_report_data():
+    donor_names = f_get_donor_names()
+    # Declare and populate lists for report data
+    total_donation, num_donation, avg_donation = [], [], []
+
+    #Lesson 4 - Use dictionary structure. Do not append data structure, only calculate in new variable
+    for name in donor_names:
+        total_donation.append(sum(dict_donors.get(name,0)))
+        num_donation.append(len(dict_donors.get(name,0)))
+        avg_donation.append(total_donation[-1]/num_donation[-1])
+
+
+    #LESSON 4
+    report = list(zip(donor_names, total_donation, num_donation, avg_donation))
+    # Sort by total donation, descending
+    report.sort(key=f_donor_key, reverse=True)
+    return report
+
+def f_print_formatted_report(report):
+    # Generate formatted report to be printed
+    # Input 'report' is expected to be a list of lists with
+    # [donor name, total donation, number of donations, average donation]
+    formatted_report = ['',
+    'Donor Name                    | Total Donation | Num Donations | Avg Donation |',
+    '-------------------------------------------------------------------------------']
+    #for donor in report:
+    #    donor_name, total, number, average = donor
+    #    formatted_report.append(f'{donor_name:<30} ${total:>14.2f}  {number:14d}  ${average:>12.2f}')
+    #formatted_report.append('')
+    #
+    #LESSON 5 - ADDED LOOP COMPREHENSION
+    print('\n'.join(formatted_report + [f'{donor_name:<30} ${total:>14.2f}  {number:14d}  ${average:>12.2f}' for donor_name, total, number, average in report]))
+    #print('\n'.join(formatted_report))
+
+def f_create_report():
+    # Generate, format, and print report data
+    report = f_generate_report_data()
+    f_print_formatted_report(report)
+
+def f_exit_program():
+    print('Exiting program...')
+    sys.exit()
+
+def f_print_error():
+
+    #Using dict switch, inputs into the switch function cannot specialize for one function. Therefore, made seperate method.
+    message1 = 'Not a valid option. Please try again!'
+    print(message1)
+
+def main():
+    # Main function, repeatedly display menu and react based on user input
+    dict_tasks = {
+        '1': f_write_thank_you,
+        '2': f_create_report,
+        '3': f_send_all_letters,
+        '4': f_exit_program
+        }
+    while True:
+        response = input(s_menu)
+        dict_tasks.get(response,f_print_error)()         #yes that extra parenthesis is necessary.......
+
+#
+#Run main

```

Comment:
lesson 05 should include exception handling -- please add exception handing to the code